### PR TITLE
test(lifecycle): add shared state substrate

### PR DESCRIPTION
**Pull request by danielmeppiel**
# add(test): provide shared lifecycle state substrate

## TL;DR

This adds the test-only foundation for the lifecycle-contract wave: realistic package/hook authoring and transitions, isolated local-Git transport, and deterministic multi-root durable-state snapshots. Later PRs can drive real commands and assert project/user-scope bytes or semantic convergence without fabricating package metadata, reimplementing lockfile serialization, reading ambient HOME, or reaching a production Git host. Production source and user-facing behavior are unchanged.

> [!IMPORTANT]
> This PR is independently mergeable foundation only. It adds no lifecycle scenarios, workflow gates, production fixes, or legacy-test deletion.

## Problem (WHY)

- [x] The existing package factory covered skills, agents, and instructions, but later lifecycle lanes also need prompts, commands, hooks, canvases, and declarative MCP/LSP inputs.
- [x] Generic filesystem snapshots preserve hashes, but they do not expose raw manifest/lock bytes, canonical deployment records, config state, hook sidecars, or compiled-output roles as one durable-state contract.
- [x] Marketplace/plugin rows need production-shaped Git URLs, but repeated inline `git config` subprocess blocks make hermetic rewrites easy to drift.
- [!] Repeating those details in each scenario PR would create parallel test vocabularies and serializer drift.

These gaps matter because ["agents pattern-match well against concrete structures"](https://agentskills.io/skill-creation/best-practices), while ["Grounding outputs in deterministic tool execution transforms probabilistic generation into verifiable action."](https://danielmeppiel.github.io/awesome-ai-native/docs/prose/) The foundation therefore composes existing production codecs and test runners instead of fabricating state.

## Approach (WHAT)

| # | Fix | Principle |
|---:|---|---|
| 1 | Extend `LocalPackageFactory` with source-only authors for every deploy category and validated MCP/LSP manifest entries. | ["Favor small, chainable primitives over monolithic frameworks."](https://danielmeppiel.github.io/awesome-ai-native/docs/prose/) |
| 2 | Add `LifecycleStateSnapshot` for raw bytes, canonical lock/ledger semantics, and workspace-contained materialized state. | ["Grounding outputs in deterministic tool execution transforms probabilistic generation into verifiable action."](https://danielmeppiel.github.io/awesome-ai-native/docs/prose/) |
| 3 | Add one owned-repository URL rewrite that writes only a bounded fixture Git config and contains Git prefix matching inside the local bare repository. | ["Favor small, chainable primitives over monolithic frameworks."](https://danielmeppiel.github.io/awesome-ai-native/docs/prose/) |
| 4 | Bind user-scope target-relative rows and selected configs through explicit, non-overlapping `LifecycleStateRoot` values. | ["Grounding outputs in deterministic tool execution transforms probabilistic generation into verifiable action."](https://danielmeppiel.github.io/awesome-ai-native/docs/prose/) |
| 5 | Author hook scripts/assets only below their descriptor-owned `.apm/hooks/<name>/` subtree. | ["Favor small, chainable primitives over monolithic frameworks."](https://danielmeppiel.github.io/awesome-ai-native/docs/prose/) |
| 6 | Replace/remove targets and APM dependencies through production validators while preserving unrelated manifest sections. | ["agents pattern-match well against concrete structures"](https://agentskills.io/skill-creation/best-practices) |
| 7 | Prove malicious paths, root collisions, missing files, non-ASCII bytes, idempotency, semantic changes, and adjacent Git prefixes. | ["do the work, run a validator (a script, a reference checklist, or a self-check), fix any issues, and repeat until validation passes."](https://agentskills.io/skill-creation/best-practices) |

## Implementation (HOW)

| File | Change |
|---|---|
| `tests/utils/local_package.py` | Extends the existing canonical factory with seven categories, hook assets, MCP/LSP declarations, and owned target/APM dependency transitions through production validators. |
| `tests/utils/lifecycle_state.py` | Adds immutable root/file/snapshot values. It delegates lock parsing and deployment rows to production owners, then binds external files only through caller-supplied target/scope roots. |
| `tests/utils/artifact_snapshot.py` | Cross-references the lifecycle-aware oracle so later lanes choose the domain snapshot instead of creating a third overlapping helper. |
| `tests/utils/local_git_repository.py` | Adds `install_url_rewrite()`: validates an owned local repository and HTTP(S) remote, derives bare/`.git` forms, writes only the scenario-bounded `GIT_CONFIG_GLOBAL`, and uses a trailing-slash local base so Git prefix suffixes cannot reach sibling fixtures. |
| `tests/integration/test_local_package_factory_contract.py` | Covers all seven deploy categories, exact source layouts/bytes, mapping and string config dependencies, transactional rejection, symlink/traversal boundaries, hook assets, and manifest transitions. |
| `tests/integration/test_lifecycle_state_snapshot_contract.py` | Covers raw and semantic state, dynamic URI rows, project plus multiple external roots, duplicate relative names, target-relative rows, missing/directory states, and every symlink/root collision boundary. |
| `tests/integration/test_local_git_repository_factory_contract.py` | Proves both URL forms with real `git ls-remote`, idempotent config, untouched HOME config, adjacent-prefix containment, and validation-before-mutation. |

### Public test API

```python
factory = LocalPackageFactory(package_root)
package = factory.create(
    "fixture",
    dependencies=({"path": "../shared", "alias": "shared"},),
    mcp_dependencies=({"name": "mcp", "registry": False,
                       "transport": "stdio", "command": "mcp"},),
    lsp_dependencies=({"name": "lsp", "command": "lsp",
                       "extensionToLanguage": {".py": "python"}},),
    targets=("copilot", "claude"),
)

factory.add_skill(package, "review", skill_text)
factory.add_agent(package, "helper", agent_text)
factory.add_prompt(package, "summarize", prompt_text)
factory.add_command(package, "check", command_text)
factory.add_instruction(package, "rules", instruction_text)
factory.add_hook(package, "lifecycle", hook_mapping)
factory.add_hook_asset(
    package, "lifecycle", PurePosixPath("scripts/check.py"),
    "print('hook')\n", executable=True,
)
factory.add_canvas(package, "inspector", extension_mjs, assets=asset_bytes)
factory.set_targets(package, ("copilot", "claude"))
factory.replace_apm_dependencies(package, dependencies)
factory.remove_apm_dependency(package, identity_equivalent_dependency)
```

The existing local-Git and command owners remain unchanged:

```python
repository_factory = LocalGitRepositoryFactory(
    repository_root,
    env=environment,
    deadline=deadline,
)
repository = repository_factory.create("fixture", source_tree=package.root)
remote_forms = repository_factory.install_url_rewrite(
    repository,
    "https://github.example.invalid/acme/fixture.git",
)

results = ApmLifecycleRunner((str(apm_binary_path),)).run_sequence(
    commands,
    expected_returncodes=expected_returncodes,
    scenario_id="required-lifecycle",
    cwd=consumer.root,
    env=environment,
)
```

Capture exact or semantic invariants from project and user roots:

```python
state = LifecycleStateSnapshot.capture(
    consumer.root,
    targets=("copilot", "claude"),
    config_paths=(PurePosixPath(".vscode/mcp.json"),),
    external_roots=(
        LifecycleStateRoot(
            root_id="claude-user",
            target="claude",
            scope="user",
            path=isolated.home / ".claude",
            config_paths=(PurePosixPath("settings.json"),),
        ),
    ),
)

assert state.manifest_bytes == expected_manifest_bytes
assert state.lockfile_bytes == expected_lockfile_bytes
assert state.deployment_records == expected_records
assert state.mcp_state_bytes == expected_mcp_state
assert state.lsp_state_bytes == expected_lsp_state
assert state.file(".agents/skills/review/SKILL.md").content == expected_skill
assert state.file("settings.json", root_id="claude-user").content == expected_user_config
assert before.semantic_bytes == after.semantic_bytes
```

## Diagram

Legend: later scenario PRs author real source and run real commands; the snapshot reads project and explicit user roots through canonical owners.

```mermaid
flowchart LR
    subgraph Author[Author source]
        LPF[LocalPackageFactory]
        SRC[Local package tree]
    end
    subgraph Execute[Execute lifecycle]
        GIT[LocalGitRepositoryFactory]
        CFG[Bounded fixture Git config]
        RUN[ApmLifecycleRunner]
        WS[Evolving workspace]
    end
    subgraph Observe[Observe durable state]
        LOCK[LockFile and DeploymentLedgerCodec]
        ROOTS[Explicit target roots]
        SNAP[LifecycleStateSnapshot]
        ASSERT[Exact or semantic assertions]
    end
    LPF --> SRC
    SRC --> GIT
    GIT --> CFG
    CFG --> GIT
    GIT --> RUN
    RUN --> WS
    WS --> LOCK
    WS --> SNAP
    ROOTS --> SNAP
    LOCK --> SNAP
    SNAP --> ASSERT
    classDef new stroke-dasharray: 5 5;
    class LPF,ROOTS,SNAP new;
```

## Trade-offs

- **Extend existing owners, not add lane-local builders.** Package, Git, runner, and snapshot responsibilities remain separate and composable.
- **Store raw bytes and canonical semantic bytes.** Raw bytes prove idempotency; canonical JSON prevents formatting-only YAML churn from appearing semantic.
- **Bound all state I/O.** Project files use the workspace root; user/target files require explicit non-overlapping `(root_id, target, scope, path)` values; URL rewrites write only scenario-bounded `GIT_CONFIG_GLOBAL`.
- **Keep source intent separate from expected deployment.** Hook assets live beneath their descriptor, while target mapping and script rewriting remain production-owned.
- **Expose substrate without migrating scenarios.** Downstream command chains consume this PR after coordinator acceptance.

## Benefits

1. Seven deploy categories, hook scripts/assets, and manifest transitions share one source-authoring vocabulary.
2. Exact state exposes manifest, lock, deployment records, MCP/LSP state, and root-qualified filesystem bytes in one capture.
3. Project and multiple user target roots can contain duplicate relative names without collisions or ambient HOME reads.
4. Bare and `.git` remotes resolve locally, while adjacent prefixes cannot resolve to sibling fixtures.
5. Invalid roots, paths, generated fields, foreign packages, and semantic round-trip drift fail before mutation.
6. The diff remains test-only: seven files changed and no `src/` modifications.

## Validation

`uv run --extra dev pytest -q tests/integration/test_local_package_factory_contract.py tests/integration/test_lifecycle_state_snapshot_contract.py tests/integration/test_local_git_repository_factory_contract.py`:

```text
..........................................                               [100%]
42 passed in 5.13s
```

<details><summary>Repository gates</summary>

`uv run --extra dev pytest -q tests/unit tests/test_console.py tests/red_team`:

```text
19292 passed, 2 skipped, 21 xfailed, 19 warnings, 88 subtests passed in 190.97s (0:03:10)
```

`uv run --extra dev pytest -p no:cacheprovider -q tests/quality` plus assertion and duplicate ratchets:

```text
37 passed in 33.72s
[+] assertion-quality ratchet clean: AQ001=5, AQ002=12
[+] exact test duplicate ratchet clean: 1062 files, 8 allowed duplicate group(s)
```

Architecture/helper contracts:

```text
108 passed in 135.81s (0:02:15)
```

Full lint mirror:

```text
All checks passed!
1545 files already formatted
Your code has been rated at 10.00/10
[+] auth-signal lint clean
[+] architecture boundary lint clean
```

Docs impact classifier:

```json
{"verdict":"no_change","confidence":"high","source":"L0","scope_pages":[]}
```

</details>

### Scenario Evidence

| # | Scenario (test-author promise) | Principle(s) | Test(s) proving it | Type |
|---:|---|---|---|---|
| 1 | A lifecycle test can author every deploy category plus MCP/LSP declarations from a real source tree. | Portability by manifest, Multi-harness support | `tests/integration/test_local_package_factory_contract.py::test_lifecycle_sources_cover_all_deploy_categories_and_config_dependencies` | integration |
| 2 | A lifecycle test can distinguish byte-only churn from a real deployment/config state change. | Governed by policy, DevX | `tests/integration/test_lifecycle_state_snapshot_contract.py::test_semantic_state_ignores_yaml_formatting_and_generated_time`<br/>`::test_semantic_state_reflects_deployment_and_config_changes` | integration |
| 3 | Malicious paths cannot make the shared helpers read or write outside their declared roots. | Secure by default | `tests/integration/test_lifecycle_state_snapshot_contract.py::test_bounded_roots_fail_closed_before_outside_reads`<br/>`::test_bounded_root_identity_and_path_collisions_are_rejected` | integration |
| 4 | A production-shaped marketplace Git URL resolves locally without touching host Git state or capturing an adjacent repository prefix. | Secure by default, Vendor-neutral, DevX | `tests/integration/test_local_git_repository_factory_contract.py::test_install_url_rewrite_is_isolated_and_resolves_bare_and_git_forms` | integration |
| 5 | Project and two user target roots can contain the same relative config name without identity or byte collisions. | Secure by default, Multi-harness support | `tests/integration/test_lifecycle_state_snapshot_contract.py::test_capture_reads_two_bounded_roots_without_relative_path_collisions` | integration |
| 6 | A hook descriptor can reference portable text/binary scripts without letting assets replace or escape its source subtree. | Secure by default, Portability by manifest | `tests/integration/test_local_package_factory_contract.py::test_hook_assets_are_bounded_source_files_with_executable_intent`<br/>`::test_hook_assets_reject_escape_replacement_and_foreign_packages` | integration |
| 7 | Lifecycle tests can widen/narrow targets and replace/remove dependencies without losing unrelated manifest sections or accepting generated fields. | Governed by policy, DevX | `tests/integration/test_local_package_factory_contract.py::test_manifest_transitions_preserve_unrelated_sections_and_are_idempotent`<br/>`::test_manifest_transitions_reject_invalid_inputs_without_partial_write` | integration |

## How to test

- [ ] Run the package/state/local-Git contracts (42 tests); expect all to pass.
- [ ] Capture project plus two user roots before/after drift; expect root-qualified bytes, semantic changes, and fail-closed escapes.
- [ ] Author a hook script and mutate targets/dependencies twice; expect canonical source paths and byte-idempotent manifests.
- [ ] Exercise bare and `.git` production URLs; expect the owned local SHA, while an adjacent prefix fails.
- [ ] Run the lint, quality, architecture, and required-test commands above; expect zero failures.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

**Comment by danielmeppiel** [comment]
## APM Review Panel: `needs_discussion`

> This tests-only foundation is well-composed and green, but its containment and target-relative behavior need explicit trust semantics before scenario PRs depend on it.

cc @sergio-sisternes-epam -- a fresh advisory pass is ready for your review.

The panel agrees that the canonical package/lock/target owners are reused correctly and that the contract tests are unusually thorough. Two findings carry the highest signal because this PR is shared substrate: the snapshot hand-walks symlink ancestors instead of composing the canonical containment helper, and target-relative deployment records disappear from file capture without an explicit failure.

Both are fixable inside the tests-only boundary. Use `ensure_path_within()` on the candidate parent so leaf symlinks can still be represented without another containment implementation, and reject target-relative rows until the API has an explicit bounded target-root input.

**Dissent.** No panelist disagreed on the risks. The CEO arbitration narrowed the remedies to test-side changes and treated the unconditional compiled-file walk as a follow-up rather than a shipping concern.

**Aligned with:** secure by default through one containment authority and fail-closed durable-state capture; OSS/community-driven through a reusable contributor substrate.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---:|---:|---:|---|
| Python Architect | 1 | 1 | 2 | Reuse canonical containment and distinguish this oracle from generic ArtifactSnapshot. |
| CLI Logging Expert | 0 | 0 | 0 | No CLI output surface changed. |
| DevX UX Expert | 0 | 3 | 3 | Tighten lookup errors and document canvas/config API asymmetries. |
| Supply Chain Security Expert | 1 | 0 | 2 | Fail closed on target-relative records; traversal and symlink coverage is otherwise strong. |
| OSS Growth Hacker | 0 | 1 | 1 | Preserve the runnable API example near the helper code. |
| Test Coverage Expert | 0 | 1 | 1 | Add legacy-lock fallback coverage and one malformed config-input case. |
| Performance Expert | 0 | 1 | 1 | Make compiled-output discovery opt-in before large scenario matrices land. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships.

### Top 5 follow-ups

1. **[Python Architect] *(blocking-severity)*** Replace the hand-written symlink-ancestor walk with canonical `ensure_path_within()` composition -- one containment authority is load-bearing for shared substrate.
2. **[Supply Chain Security Expert] *(blocking-severity)*** Reject target-relative deployment records explicitly -- silent omission can make a later cleanup assertion trust an incomplete snapshot.
3. **[Test Coverage Expert]** Add a legacy `apm.lock` fallback contract -- the new branch currently has no direct proof.
4. **[DevX UX Expert]** Make untracked `file()` lookups fail clearly and document `add_canvas()` returning a bundle directory -- this keeps downstream failures actionable.
5. **[Performance Expert]** Avoid unconditional full-tree compiled-output discovery -- future `apm_modules` fixtures should pay for the scan only when requested.

### Architecture

```mermaid
classDiagram
    direction LR
    class LocalPackageFactory {
      <<Factory>>
      +create(...) LocalPackage
      +add_skill(...) Path
      +add_agent(...) Path
      +add_prompt(...) Path
      +add_command(...) Path
      +add_hook(...) Path
      +add_canvas(...) Path
    }
    class LifecycleStateSnapshot {
      <<Facade>>
      +capture(...) LifecycleStateSnapshot
      +file(relative_path) LifecycleFileState
    }
    class LifecycleFileState {
      <<ValueObject>>
      +relative_path str
      +kind LifecycleFileKind
      +roles frozenset
      +content bytes
      +sha256 str
    }
    class LockFile { <<CanonicalOwner>> }
    class DeploymentLedgerCodec { <<CanonicalOwner>> }
    class PathSecurity { <<CanonicalOwner>> }
    class ArtifactSnapshot { <<ValueObject>> }

    LifecycleStateSnapshot ..> LockFile : parses
    LifecycleStateSnapshot ..> DeploymentLedgerCodec : orders rows
    LifecycleStateSnapshot ..> PathSecurity : contains reads
    LifecycleStateSnapshot *-- LifecycleFileState : captures
    LifecycleStateSnapshot ..> ArtifactSnapshot : complementary
```

```mermaid
flowchart TD
    A[LocalPackageFactory] --> B[Real package source]
    B --> C[LocalGitRepositoryFactory]
    C --> D[ApmLifecycleRunner]
    D --> E[Evolving workspace]
    E --> F[LockFile and DeploymentLedgerCodec]
    E --> G[LifecycleStateSnapshot]
    F --> G
    G --> H[Exact byte assertions]
    G --> I[Semantic convergence assertions]
```

### Recommendation

Fold the two trust-boundary fixes before downstream scenario branches consume the API. Then keep the legacy-lock proof and small DevX/documentation improvements in this PR; treat scaling work as a bounded follow-up unless the new opt-in can be added without API churn.

---

<details>
<summary>Full per-persona findings</summary>

#### Python Architect

- **[blocking]** A hand-written `_reject_symlink_ancestors()` creates another containment authority in `tests/utils/lifecycle_state.py`.
  The leaf-symlink distinction is valid, but parent containment can compose production `ensure_path_within()` instead of maintaining a separate walk.
  *Suggested:* validate `path.parent` against the workspace root and keep leaf classification in `_capture_file()`.
- **[recommended]** `ArtifactSnapshot.capture()` and `LifecycleStateSnapshot.capture()` need reciprocal docstring guidance.
  The abstractions are complementary, but future authors need a clear choice.
- **[nit]** Importing private `_APM_HOOKS_SIDECAR` is intentional reuse but remains a private coupling.
- **[nit]** The SHA-256 expression duplicates the generic snapshot helper; extraction is not justified at two call sites.

#### CLI Logging Expert

No findings.

#### DevX UX Expert

- **[recommended]** `LifecycleStateSnapshot.file()` returns `None` for an untracked or misspelled path.
  *Suggested:* raise `KeyError` with the tracked paths, while preserving `kind="missing"` for tracked-but-absent files.
- **[recommended]** `add_canvas()` returns a directory while sibling `add_*` methods return files.
  *Suggested:* document the bundle-directory return in the method docstring.
- **[recommended]** `add_canvas()` calls private `CanvasIntegrator._validate_canvas_name`.
  This is the only production canvas-name owner today; keep the coupling explicit in documentation rather than duplicate validation.
- **[nit]** `mcp_state_json` and `lsp_state_json` are bytes despite the suffix.
- **[nit]** `ConfigDependencyInput` duplicates the shape of `DependencyInput`.
- **[nit]** `add_command()` should state that command sources use `.apm/prompts/*.prompt.md`.

#### Supply Chain Security Expert

- **[blocking]** `TARGET_RELATIVE` deployment records are retained semantically but omitted from file capture without an error.
  *Proof (missing at integration-with-fixtures):* `tests/integration/test_lifecycle_state_snapshot_contract.py::test_capture_rejects_target_relative_deployment_without_bounded_root` -- proves the oracle cannot silently claim complete file state.
  *Suggested:* reject the unsupported locator kind until a bounded target-root API exists.
- **[nit]** Reject Windows drive syntax consistently with sibling source-path validation.
- **[nit]** Convert `NotADirectoryError` into an actionable path error.

#### OSS Growth Hacker

- **[recommended]** The PR body's runnable API example should survive near the helper.
  *Suggested:* add concise module/class examples rather than requiring PR archaeology.
- **[nit]** A contributor-guide pointer can follow after the lifecycle wave stabilizes.

#### Auth Expert -- inactive

Only tests changed; no token, credential, host, clone, header, or AuthResolver surface changed.

#### Doc Writer -- inactive

Only tests changed and docs impact classified as high-confidence `no_change`.

#### Test Coverage Expert

- **[recommended]** The legacy `apm.lock` fallback branch has no direct contract test.
  *Proof (missing at integration-with-fixtures):* `tests/integration/test_lifecycle_state_snapshot_contract.py::test_capture_reads_legacy_apm_lock_when_canonical_lockfile_is_absent`.
- **[nit]** Non-string/non-mapping MCP/LSP dependency inputs do not exercise the new TypeError branch.

#### Performance Expert

- **[recommended]** `_compiled_paths()` walks the full workspace on every capture.
  *Suggested:* make compiled-output discovery opt-in so semantic-only captures stay proportional to selected durable state.
- **[nit]** Defer a scaling guard until realistic lifecycle matrices establish a useful fixture size and budget.

</details>

<sub>This panel is advisory. It does not block merge. Re-apply the
`panel-review` label after addressing feedback to re-run.</sub>

File: tests/utils/local_package.py [hunk 1]
```diff
@@ -306,6 +422,47 @@ def _validate_dependencies(
             validated.append(source_entry)
         return validated
 
+    @staticmethod
+    def _validate_config_dependencies(
+        dependencies: Sequence[ConfigDependencyInput],
+        *,
+        kind: str,
+    ) -> list[str | dict[str, object]]:
+        model = MCPDependency if kind == "MCP" else LSPDependency
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply tests/utils/local_package.py]: Fixed in 78a516c92a: the dispatch is now explicit for MCP and LSP, raises on unknown kinds, and has a regression assertion for the typo path.

**Comment by danielmeppiel** [comment]
## APM Review Panel: `ship_with_followups`

> The final tests-only foundation composes production owners cleanly and has no blocking-severity findings across the panel; one security panel return failed schema validation, with an independent security review covering the exact head.

cc @sergio-sisternes-epam -- a fresh advisory pass is ready for your review.

The panel converges on shipping the shared foundation now. The multi-root state oracle, hook assets, manifest transitions, and isolated Git transport remain test-only and route target, scope, dependency identity, lockfile, and path decisions through production owners. Exact-head evidence is strong: 41 focused, 107 helper/architecture, 37 quality, and 19,292 required tests pass.

The supply-chain panelist failed its return schema twice. That is a panel-process defect, not a clean security opinion; an independent supply-chain specialist reviewed this exact diff/head and found no security gap, corroborated by the filesystem and real-Git contracts.

**Dissent.** No substantive disagreement. Recommended items are documentation/coverage polish or forward-looking test-runtime concerns, not correctness faults in the accepted API.

**Aligned with:** secure by default through explicit non-overlapping roots and validation-before-write; multi-harness support through target/scope-qualified state; pragmatic contributor UX through one reusable fixture vocabulary.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---:|---:|---:|---|
| Python Architect | 0 | 2 | 1 | Canonical composition is sound; private production couplings deserve future public owners. |
| CLI Logging Expert | 0 | 0 | 0 | No CLI output/logging surface changed. |
| DevX UX Expert | 0 | 3 | 3 | API is usable; document intentional asymmetries and keep accepted names stable. |
| Supply Chain Security Expert | 0 | 0 | 0 | Schema failure placeholder; independent exact-head security review is clean. |
| OSS Growth Hacker | 0 | 0 | 1 | Internal discoverability is adequate; a helper index is optional. |
| Auth Expert | 0 | 0 | 0 | Test-only file transport does not alter production credential resolution. |
| Test Coverage Expert | 0 | 0 | 1 | Required scenarios pass; only synthetic round-trip error branches lack direct proof. |
| Performance Expert | 0 | 1 | 2 | Current scale is healthy; full-workspace compiled discovery is a future test-runtime concern. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships.

### Top 5 follow-ups

1. **[Test Coverage Expert]** Add a synthetic MCP/LSP YAML round-trip drift regression -- it is the only concrete missing branch proof.
2. **[Supply Chain Security Process]** Repair the panelist return-schema failure -- independent review covers this PR, but the panel process should not degrade silently.
3. **[Python Architect]** Track public production owners for hook sidecar/canvas validation and lockfile fallback -- this test-only PR must not widen into `src/`.
4. **[DevX UX Expert]** Document canvas bulk-only assets and discard-style dependency removal -- preserve accepted API names.
5. **[Performance Expert]** Reassess whole-workspace compiled-output discovery when scenario fixtures materialize large module trees.

### Architecture

```mermaid
classDiagram
    direction LR
    class LocalPackageFactory {
      <<Factory>>
      +create(...) LocalPackage
      +add_hook_asset(...) Path
      +set_targets(...) None
      +replace_apm_dependencies(...) None
      +remove_apm_dependency(...) bool
    }
    class LifecycleStateRoot {
      <<ValueObject>>
      +root_id str
      +target str
      +scope str
      +path Path
    }
    class LifecycleStateSnapshot {
      <<ValueObject>>
      +capture(..., external_roots)$ LifecycleStateSnapshot
      +file(relative_path, root_id) LifecycleFileState
    }
    class LocalGitRepositoryFactory {
      <<Factory>>
      +install_url_rewrite(repository, remote_url) tuple
    }
    class LockFile { <<ProductionOwner>> }
    class DeploymentLedgerCodec { <<ProductionOwner>> }
    class DependencyReference { <<ProductionOwner>> }
    LifecycleStateSnapshot o-- LifecycleStateRoot : bounded by
    LifecycleStateSnapshot ..> LockFile : parses
    LifecycleStateSnapshot ..> DeploymentLedgerCodec : reads rows
    LocalPackageFactory ..> DependencyReference : validates identity
```

```mermaid
flowchart TD
    A[LocalPackageFactory source tree] --> B[LocalGitRepositoryFactory]
    B --> C[Bounded Git config]
    C --> D[ApmLifecycleRunner]
    D --> E[Project state root]
    D --> F[Explicit user target roots]
    E --> G[LifecycleStateSnapshot]
    F --> G
    G --> H[Exact bytes and semantic state]
```

### Recommendation

Ship the accepted shared foundation after the remaining in-scope coverage/docstring polish is folded. Keep production public-owner cleanup and scaling work as explicit follow-ups; neither belongs in this tests-only PR.

---

<details>
<summary>Full per-persona findings</summary>

#### Python Architect

- **[recommended]** Tests consume private hook-sidecar and canvas-validator symbols.
  The single-owner intent is correct; a later production PR can expose public names without widening this test-only change.
- **[recommended]** Lockfile current/legacy fallback still lacks one production resolver owner.
  Track the production consolidation separately.
- **[nit]** Dependency/config round-trip validation shares a similar algorithm at only two call sites.

#### CLI Logging Expert

No findings.

#### DevX UX Expert

- **[recommended]** The accepted Git helper name differs from the reviewer's preferred verb-first alternative.
  Coordinator acceptance makes stability higher value than a cosmetic rename.
- **[recommended]** Canvas assets are bulk-only while hook assets are incremental.
  *Suggested:* document the intentional distinction.
- **[recommended]** Add a user-scope external-root worked example.
  The updated PR body now includes this example.
- **[nit]** The URL rewrite returns an anonymous two-tuple.
- **[nit]** Document discard-style `remove_apm_dependency`.
- **[nit]** Harmonize prompt/command source-layout docstrings.

#### Supply Chain Security Expert

The panelist failed schema validation twice. An independent exact-head supply-chain review found no security gaps in bounded roots, hook assets, manifest transitions, or Git URL rewriting.

#### OSS Growth Hacker

- **[nit]** A tests/utils helper index could improve future internal discovery, but current docstrings, contract names, and PR examples are sufficient.

#### Auth Expert

No findings.

#### Doc Writer -- inactive

Tests-only diff; docs impact remains high-confidence `no_change`.

#### Test Coverage Expert

- **[nit]** MCP/LSP YAML round-trip semantic-drift error branches have no direct synthetic regression.
  *Proof (missing at unit):* `tests/integration/test_local_package_factory_contract.py::test_config_dependency_round_trip_guard_fires_on_semantic_drift`.

#### Performance Expert

- **[recommended]** Compiled-output discovery walks the full workspace when file-backed targets are requested.
  This is required for distributed AGENTS.md discovery today; revisit when large scenario fixtures establish a measurable cost.
- **[nit]** External-root overlap validation is O(k^2) with tiny bounded `k`.
- **[nit]** Git config subprocess count is bounded and idempotent; no action.

</details>

<sub>This panel is advisory. It does not block merge. Re-apply the
`panel-review` label after addressing feedback to re-run.</sub>
